evgMrmApp/src/evgMrm.cpp

```diff
@@ -194,34 +293,139 @@ evgMrm::resetMxc(bool reset) {
 }
 
 void
-evgMrm::isr(void* arg) {
+evgMrm::isr_pci(void* arg) {
+    evgMrm *evg = (evgMrm*)(arg);
+
+    // Call to the generic implementation
+    evg->isr(arg);
+
+    /**
+     * On PCI veriant of EVG the interrupts get disabled in kernel (by uio_mrf module) since IRQ task is completed here (in userspace).
+     * Interrupts must therfore be renabled here.
+     *
+     * Change by: tslejko
+     * Reason: cPCI support
+     */
+    if(devPCIEnableInterrupt(evg->m_pciDevice)) {
+        printf("PCI: Failed to enable interrupt\n");
+        return;
+    }
+}
+
+void
+evgMrm::isr_vme(void* arg) {
     evgMrm *evg = (evgMrm*)(arg);
 
     epicsUInt32 flags = READ32(evg->m_pReg, IrqFlag);
     epicsUInt32 enable = READ32(evg->m_pReg, IrqEnable);
     epicsUInt32 active = flags & enable;
-    
+
+    // This skips extra work with a shared interrupt.
     if(!active)
-        return;
+      return;
+
+    // Call to the generic implementation
+    evg->isr(arg);
+}
+
+void
+evgMrm::isr(void* arg) {
+    evgMrm *evg = (evgMrm*)(arg);
+
+    epicsUInt32 flags = READ32(evg->m_pReg, IrqFlag);
+    epicsUInt32 enable = READ32(evg->m_pReg, IrqEnable);
+    epicsUInt32 active = flags & enable;
+
+    #ifdef vxWorks
+    /* actually: if isr runs in kernel mode */
+
+        /*
+         * For IFC1210 board this is useless
+         * and for SwissFEL_TIM it is dangerous, since
+         * in the unlikely event of queuing more than 2 IRQs the system
+         * will stop dropping them. For SwissFEL timing system this is
+         * unacceptable.
+         * Furthermore it is not thread safe!!!
+         * A race condition has been observed where
+         * x_queued is changed back to 0 without re-enabling interrupts.         *
+         * For PSI, this is now handled in the IRQ thread, avg ISR time is
+         * around ~10us
+         */
```

Comment:
FYI I suspect the behavior described by @ts is a side effect of a race condition with the IRQEnable register, and not with the *_queued flags.

That said, for SMP systems it may be reasonable optimization to do this work directly in the ISR since devLib2 gives each ISR it's own thread.  Care must be taken as this as ISR threads are epicsThreadPriorityMax-1 (99) which the high priority CB thread is epicsThreadPriorityHigh (90).